Intersect.Client/Interface/Game/Typewriting/Typewriter.cs

```diff
@@ -0,0 +1,145 @@
+        }
+
+        public void Initialize(List<Label> labels)
+        {
+            _nextUpdateTime = Timing.Global.Milliseconds;
+            _labels = labels;
+            _lines = _labels.Select(l => l.Text).ToArray();
+            _labels.ForEach(l => l.SetText(string.Empty)); // Clear the text from the labels. The writer is the captain now
+            _lineIndex = 0;
+            _charIndex = 0;
+            _lastChar = null;
+            Done = false;
+        }
+
+        public void Write(string voice)
+        {
+            if (Done)
+            {
+                return;
+            }
+
+            if (string.IsNullOrWhiteSpace(_currentLine) || _currentLabel == default)
+            {
+                End();
+                return;
+            }
+
+            if (Timing.Global.Milliseconds < _nextUpdateTime)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(voice) && _charIndex % ClientConfiguration.Instance.TypewriterSoundFrequency == 0)
+            {
+                Audio.StopAllGameSoundsOf(ClientConfiguration.Instance.TypewriterSounds.ToArray());
+                Audio.AddGameSound(voice ?? string.Empty, false);
+            }
+
+            _charIndex++;
+            if (_charIndex >= _currentLine.Length)
+            {
+                _currentLabel?.SetText(_currentLine);
+                NewLine();
+                return;
+            }
+
+            _lastChar = _currentLine[_charIndex - 1];
+
+            var written = _currentLine?.Substring(0, _charIndex) ?? string.Empty;
+            _currentLabel?.SetText(written);
+
+            _nextUpdateTime = Timing.Global.Milliseconds + GetTypingDelayFor(_currentLine.ElementAtOrDefault(_charIndex), _lastChar.Value);
+        }
+
+        private static long GetTypingDelayFor(char currentChar, char? lastChar)
+        {
+            if (lastChar == null)
+            {
+                return _typingSpeed;
+            }
+
+            var lastCharVal = lastChar.GetValueOrDefault();
+            // Allows things like ellipses
+            if (currentChar == lastCharVal)
+            {
+                return _typingSpeed;
+            }
+            if (_fullstopChars.Contains(lastCharVal))
+            {
+                return _fullStopSpeed;
+            }
+            if (_partialstopChars.Contains(lastCharVal))
+            {
+                return _partialStopSpeed;
+            }
+
+            return _typingSpeed;
+        }
+
+        public void End()
+        {
+            if (Done || (_lines?.Length ?? 0) == 0)
+            {
+                return;
+            }
+
+            for (var i = 0; i < _lines.Length; i++)
+            {
+                if (i >= _labels.Count)
+                {
+                    continue;
+                }
+
+                _labels[i].SetText(_lines[i]);
+            }
+
+            Done = true;
+            DoneAt = Timing.Global.Milliseconds;
+        }
+    }
+}
```

Comment:
**PascalCase convention for public properties:** 
 `Done` and `DoneAt`
**should be renamed to:**
 `IsDone` and `DoneAtMilliseconds`

**camelCase convention for private fields:**
__fullstopChars, _fullStopSpeed, _partialstopChars, _partialStopSpeed, and _typingSpeed_ 
**should be renamed to:** 
_FullStopChars, FullStopSpeed, PartialStopChars, PartialStopSpeed, and TypingSpeed._

**camelCase convention for private fields and local variables:**
__labels, _lines, _lineIndex, _charIndex, _lastChar, and _nextUpdateTime_ 
**should be renamed to:**
_labels, lines, lineIndex, charIndex, lastChar, and nextUpdateTime_

Reply:
I'm fine with the first two, but would have to hear what @lodicolo wants for the latter ones since the underscore convention was his suggestion

Reply:
well, these were just suggestions, the maintainer has the last word on it 😄 i believe that we actually use "m" instead of "_" as prefix in other files but i could be very wrong on this one too 😆